File: atosim/org.palladiosimulator.blockchainsystems.threesim/src/main/java/org/palladiosimulator/blockchainsystems/threesim/metrics/calculators/RaceAttackSuccessCalculator.java
package org.palladiosimulator.blockchainsystems.threesim.metrics.calculators;

import org.palladiosimulator.blockchainsystems.threesim.metrics.RaceAttackSuccess;
import org.palladiosimulator.blockchainsystems.threesim.metrics.abstractions.AverageOutputMetric;
import org.palladiosimulator.blockchainsystems.threesim.metrics.abstractions.AverageOutputMetricImpl;
import java.util.List;

public class RaceAttackSuccessCalculator {
    private final boolean success;
    public RaceAttackSuccessCalculator(boolean success) { this.success = success; }
    public RaceAttackSuccess calculate() { return new RaceAttackSuccess(success); }

    public static AverageOutputMetric calculateAverage(List<RaceAttackSuccess> values) {
        double avg = values.stream().mapToDouble(v -> v.getValue()).average().orElse(0.0);
        double variance = values.stream().mapToDouble(v -> (v.getValue() - avg) * (v.getValue() - avg)).average().orElse(0.0);
        double std = Math.sqrt(variance);
        Double cv = avg == 0.0 ? null : std / avg;
        return new AverageOutputMetricImpl(RaceAttackSuccess.NAME, avg, null, std, cv);
    }
}

File: atosim/org.palladiosimulator.blockchainsystems.threesim/src/main/kotlin/org/palladiosimulator/blockchainsystems/threesim/metrics/calculators/RaceAttackSuccessCalculator.kt
package org.palladiosimulator.blockchainsystems.threesim.metrics.calculators

import org.palladiosimulator.blockchainsystems.threesim.metrics.RaceAttackSuccess
import org.palladiosimulator.blockchainsystems.threesim.metrics.abstractions.AverageOutputMetric
import org.palladiosimulator.blockchainsystems.threesim.metrics.abstractions.AverageOutputMetricImpl
import kotlin.math.sqrt

class RaceAttackSuccessCalculator(
    private val success: Boolean
) {

    fun calculate(): RaceAttackSuccess {
        return RaceAttackSuccess(success)
    }

    companion object {

        fun calculateAverage(values: List<RaceAttackSuccess>): AverageOutputMetric {

            val avg = values.map { it.value }.average()

            val variance = values
                .map { (it.value - avg) * (it.value - avg) }
                .average()

            val std = sqrt(variance)

            val cv = if (avg == 0.0) null else std / avg

            return AverageOutputMetricImpl(
                name = RaceAttackSuccess.NAME,
                average = avg,
                unit = null,
                standardDeviation = std,
                coefficientOfVariation = cv
            )
        }
    }
}